Add project/build filters

agjohnson opened the pull request:
Most of these filters are currently only used in views when the new
templates are enabled. The project dashboard view context data was
replaced with a filter, which does enable sorting on the dashboard view
using request arguments (like `?sort=built_last`), but this isn't
exposed in the UI.

agjohnson commented:
I found a missing commit that was supposed to be part of this branch. Going back to WIP.

agjohnson commented:
Ready for review!

humitos commented on readthedocs/projects/filters.py:
```diff
@@ -0,0 +1,158 @@
+log = logging.getLogger(__name__)
+
+
+class SortOrderingFilter(OrderingFilter):
```
I didn't find _where_ you use this custom filter.

agjohnson replied:
Hrm, then I might have busted something in all of the merges. Let me check

humitos commented on readthedocs/projects/filters.py:
```diff
@@ -0,0 +1,158 @@
+    """
+    Version list sort ordering django_filters filter.
+
+    Django-filter is highly opionated, and the default model filters do not work
+    well with empty/null values in the filter choices. In our case, empty/null
+    values are used for a default query. So, to make this work, we will use a
+    custom filter, instead of an automated model filter.
```
By reading this comment, I'm not sure to understand what's the problem that this filter solves 😅 --I only get that it does "something" with null values, but not sure what fixes from the default filter.

agjohnson replied:
:+1: I can expand on this

humitos commented on readthedocs/projects/views/private.py:
```diff
@@ -100,6 +101,13 @@ def get_context_data(self, **kwargs):
         # Set the default search to search files instead of projects
         context['type'] = 'file'
+
+        filter = ProjectListFilterSet(self.request.GET, queryset=self.get_queryset())
+        context['filter'] = filter
+        context['project_list'] = filter.qs
+        # Alternatively, dynamically override super()-derived `project_list` context_data
+        # context[self.get_context_object_name(filter.qs)] = filter.qs
```
Shouldn't this be behind `if settings.RTD_EXT_THEME_ENABLED`?

agjohnson replied:
I think you're right. I assumed that I wrote this to replace existing filtering functionality on `project_list`, but digging through code I don't see any similar filter mechanism.

humitos reviewed (approved):
The code looks good to me. However, I haven't QAed this yet because it seems it's not ready from the UI. This is not used yet, so I think it's fine to merge and leave it ready for the new templates to make usage of it.

agjohnson commented:
I fixed some goofs, but that probably means I should check the application again with the new templates because I'm sure I've changed something incorrectly. I will test this today.

ericholscher commented on readthedocs/builds/filters.py:
```diff
@@ -0,0 +1,43 @@
+        (STATE_ACTIVE, _('Active')),
+        (STATE_SUCCESS, _('Build finished')),
+        (STATE_FAILED, _('Build failed')),
+    )
```
Should these live in a constants file somewhere? Seems useful outside of just the filter.

agjohnson replied:
These are very particular to the filter UI, and I wouldn't expect to reuse them.

ericholscher commented on readthedocs/builds/views.py:
```diff
@@ -97,7 +95,13 @@ def get_context_data(self, **kwargs):  # pylint: disable=arguments-differ
-        context['build_qs'] = self.get_queryset()
+
+        builds = self.get_queryset()
+        if settings.RTD_EXT_THEME_ENABLED:
```
I think we can just always run it through the filter, no need for the setting. It just won't get filtered, if the user doesn't have the GET args.

agjohnson replied:
That was what I originally had here, but updated on feedback already. I think the conditional is fine, we'll test this in the beta on production data.

ericholscher commented on readthedocs/projects/filters.py:
```diff
@@ -0,0 +1,163 @@
+        # doesn't work with a standard model filter.
+        if value is None:
+            value = ['relevance']
+        return qs.annotate(
```
Don't we only want to annotate the value that we're actually filtering? Seems like this will add query overhead.

agjohnson replied:
Yeah, that's an easy optimization to make. I'll update this

ericholscher commented on readthedocs/projects/filters.py:
```diff
@@ -0,0 +1,163 @@
+    Django-filter is highly opionated, and the default model filters do not work
+    well with empty/null values in the filter choices. In our case, empty/null
+    values are used for a default query. So, to make this work, we will use a
+    custom filter, instead of an automated model filter.
```
Seems like that's what the null_choice is? https://django-filter.readthedocs.io/en/stable/guide/tips.html#solution-2-using-choicefilter-s-null-choice

agjohnson replied:
I wasn't able to accomplish what I was looking for with null_choice

ericholscher reviewed (commented):
This looks like a good start. I think we should just always use the filter, and that will let us figure out if there's performance issues with it. 

Seems like this also needs some tests. I'm a little worried about performance here, so it might be good to make sure the queries we're generating aren't going to kill the DB, especially around the Build & Version tables.

agjohnson commented:
> Seems like this also needs some tests.

Perhaps. I'm probably not the best person to write these tests however. It would probably be better if someone else wants to dive into test and/or we revisit this later. My major concern is continuing to block this work, it's holding up my progress on templates.

ericholscher reviewed (approved):
I'm fine merging and adding a TODO somewhere to test this code, as long as we actually get back to it :)

I'd love to have you working w/ @humitos as a small team, with him doing backend & API changes needed. I'd generally like to see us doing this more, but we all have close to the same skillsets which makes it hard to find space for it.

agjohnson commented:
Yup, that's the workflow I was describing.

I've already pushed off API changes, which are easier to collaborate on, though differences in priority slow progress a good deal.

Changes like this PR are hard, and require experimentation on multiple repositories to nail down the UI. This process would have to be more of a pairing process, as there was experimentation on multiple repositories to get this UI correct. It would have been an awful amount of back and forth to collaborate on this work. This is where I'm looking for process improvements. We will want to figure out how to split up backend and frontend development into digestible work.

agjohnson commented:
Moved test coverage to #8238

agjohnson commented:
I have tested these changes and they look correct. Build/version/project filters are all working on new templates. I squeezed a small fix for unpacking the dictionary, I didn't get a chance to test the previous commit because my dev environment was broken again. Looks good now however.
